--- ../../alsa-kernel/isa/sb/sb16_csp.c	2007-04-26 14:02:53.000000000 +0200
+++ sb16_csp.c	2007-04-26 14:02:54.000000000 +0200
@@ -723,8 +723,12 @@
 #ifdef CONFIG_SND_SB16_CSP_FIRMWARE_IN_KERNEL
 		program = &snd_sb_csp_static_programs[index];
 #else
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 5, 0)
 		int err = request_firmware(&program, names[index],
-				       p->chip->card->dev);
+					   p->chip->card->dev);
+#else
+		int err = request_firmware(&program, names[index], "sb16_csp");
+#endif
 		if (err < 0)
 			return err;
 #endif
