--- seq_clientmgr.c	2003-05-30 15:37:39.000000000 +0200
+++ seq_clientmgr.c.old	2003-05-30 15:38:16.000000000 +0200
@@ -1,3 +1,4 @@
+#define __NO_VERSION__
 /*
  *  ALSA sequencer Client Manager
  *  Copyright (c) 1998-2001 by Frank van de Pol <fvdpol@home.nl>
@@ -2454,7 +2455,9 @@
 
 static struct file_operations snd_seq_f_ops =
 {
+#ifndef LINUX_2_2
 	.owner =	THIS_MODULE,
+#endif
 	.read =		snd_seq_read,
 	.write =	snd_seq_write,
 	.open =		snd_seq_open,
