--- ../alsa-kernel/core/init.c	2012-04-05 17:55:31.000000000 +0200
+++ init.c	2012-07-19 14:04:35.000000000 +0200
@@ -1,3 +1,5 @@
+#define __NO_VERSION__
+#include "adriver.h"
 /*
  *  Initialization routines
  *  Copyright (c) by Jaroslav Kysela <perex@perex.cz>
@@ -298,11 +300,19 @@
 	return POLLERR | POLLNVAL;
 }
 
+#ifdef CONFIG_SND_HAVE_NEW_IOCTL
 static long snd_disconnect_ioctl(struct file *file,
 				 unsigned int cmd, unsigned long arg)
 {
 	return -ENODEV;
 }
+#else
+static int snd_disconnect_ioctl_old(struct inode *inode, struct file *file,
+				    unsigned int cmd, unsigned long arg)
+{
+	return -ENODEV;
+}
+#endif
 
 static int snd_disconnect_mmap(struct file *file, struct vm_area_struct *vma)
 {
@@ -322,10 +332,14 @@
 	.write =	snd_disconnect_write,
 	.release =	snd_disconnect_release,
 	.poll =		snd_disconnect_poll,
+#ifdef CONFIG_SND_HAVE_NEW_IOCTL
 	.unlocked_ioctl = snd_disconnect_ioctl,
 #ifdef CONFIG_COMPAT
 	.compat_ioctl = snd_disconnect_ioctl,
 #endif
+#else
+	.ioctl =	snd_disconnect_ioctl_old,
+#endif
 	.mmap =		snd_disconnect_mmap,
 	.fasync =	snd_disconnect_fasync
 };
@@ -394,10 +408,12 @@
 		snd_printk(KERN_ERR "not all devices for card %i can be disconnected\n", card->number);
 
 	snd_info_card_disconnect(card);
+#ifndef CONFIG_SYSFS_DEPRECATED
 	if (card->card_dev) {
 		device_unregister(card->card_dev);
 		card->card_dev = NULL;
 	}
+#endif
 #ifdef CONFIG_PM
 	wake_up(&card->power_sleep);
 #endif
@@ -600,6 +616,7 @@
 }
 EXPORT_SYMBOL(snd_card_set_id);
 
+#ifndef CONFIG_SYSFS_DEPRECATED
 static ssize_t
 card_id_show_attr(struct device *dev,
 		  struct device_attribute *attr, char *buf)
@@ -651,6 +668,7 @@
 
 static struct device_attribute card_number_attrs =
 	__ATTR(number, S_IRUGO, card_number_show_attr, NULL);
+#endif /* CONFIG_SYSFS_DEPRECATED */
 
 /**
  *  snd_card_register - register the soundcard
@@ -669,15 +687,27 @@
 
 	if (snd_BUG_ON(!card))
 		return -EINVAL;
-
+#ifndef CONFIG_SYSFS_DEPRECATED
 	if (!card->card_dev) {
+#ifdef CONFIG_SND_HAS_DEVICE_CREATE_DRVDATA
+		card->card_dev = device_create_drvdata(sound_class, card->dev,
+						       MKDEV(0, 0), card,
+						       "card%i", card->number);
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
 		card->card_dev = device_create(sound_class, card->dev,
 					       MKDEV(0, 0), card,
 					       "card%i", card->number);
+#else
+		card->card_dev = device_create(sound_class, card->dev,
+					       MKDEV(0, 0),
+					       "card%i", card->number);
+		if (card->card_dev && !IS_ERR(card->card_dev))
+			dev_set_drvdata(card->card_dev, card);
+#endif
 		if (IS_ERR(card->card_dev))
 			card->card_dev = NULL;
 	}
-
+#endif
 	if ((err = snd_device_register_all(card)) < 0)
 		return err;
 	mutex_lock(&snd_card_mutex);
@@ -705,6 +735,7 @@
 	if (snd_mixer_oss_notify_callback)
 		snd_mixer_oss_notify_callback(card, SND_MIXER_OSS_NOTIFY_REGISTER);
 #endif
+#ifndef CONFIG_SYSFS_DEPRECATED
 	if (card->card_dev) {
 		err = device_create_file(card->card_dev, &card_id_attrs);
 		if (err < 0)
@@ -713,7 +744,7 @@
 		if (err < 0)
 			return err;
 	}
-
+#endif
 	return 0;
 }
 
@@ -979,3 +1010,16 @@
 
 EXPORT_SYMBOL(snd_power_wait);
 #endif /* CONFIG_PM */
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 30)
+/* just for old ABI compatibility */
+struct snd_card *snd_card_new(int idx, const char *id,
+			      struct module *module, int extra_size)
+{
+	struct snd_card *card;
+	if (snd_card_create(idx, id, module, extra_size, &card) < 0)
+		return NULL;
+	return card;
+}
+EXPORT_SYMBOL(snd_card_new);
+#endif
