--- memalloc.c.old	2004-03-02 11:09:58.000000000 +0100
+++ memalloc.c	2004-03-02 11:09:17.000000000 +0100
@@ -1,3 +1,4 @@
+#include "memalloc.inc"
 /*
  *  Copyright (c) by Jaroslav Kysela <perex@suse.cz>
  *                   Takashi Iwai <tiwai@suse.de>
@@ -461,7 +462,7 @@
 #endif
 #ifdef CONFIG_PCI
 	case SNDRV_DMA_TYPE_PCI:
-		dmab->area = snd_malloc_dev_pages(&dev->dev.pci->dev, size, &dmab->addr);
+		dmab->area = snd_malloc_dev_pages(snd_kdevice_pci(dev->dev.pci), size, &dmab->addr);
 		break;
 #endif
 	case SNDRV_DMA_TYPE_DEV:
@@ -522,7 +523,7 @@
 #endif
 #ifdef CONFIG_PCI
 	case SNDRV_DMA_TYPE_PCI:
-		dmab->area = snd_malloc_dev_pages_fallback(&dev->dev.pci->dev, size, &dmab->addr, &dmab->bytes);
+		dmab->area = snd_malloc_dev_pages_fallback(snd_kdevice_pci(dev->dev.pci), size, &dmab->addr, &dmab->bytes);
 		break;
 #endif
 	case SNDRV_DMA_TYPE_DEV:
@@ -569,7 +570,7 @@
 #endif
 #ifdef CONFIG_PCI
 	case SNDRV_DMA_TYPE_PCI:
-		snd_free_dev_pages(&dev->dev.pci->dev, dmab->bytes, dmab->area, dmab->addr);
+		snd_free_dev_pages(snd_kdevice_pci(dev->dev.pci), dmab->bytes, dmab->area, dmab->addr);
 		break;
 #endif
 	case SNDRV_DMA_TYPE_DEV:
@@ -724,6 +725,7 @@
 		kfree(mem);
 	}
 	up(&list_mutex);
+	snd_kdevice_free_all();
 }
 
 
@@ -771,7 +773,7 @@
 {
 	memset(dev, 0, sizeof(*dev));
 	dev->type = SNDRV_DMA_TYPE_DEV;
-	dev->dev.dev = &pci->dev;
+	dev->dev.dev = snd_kdevice_pci(pci);
 	dev->id = id;
 }
 
@@ -873,7 +875,11 @@
 			if (mem->dev.dev.dev) {
 				len += sprintf(page + len, "%s [%s]",
 					       mem->dev.type == SNDRV_DMA_TYPE_DEV_SG ? "DEV" : "DEV-SG",
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 5, 0)
 					       mem->dev.dev.dev->bus_id);
+#else
+					       snd_kdevice_busid(mem->dev.dev.dev));
+#endif
 			}
 			break;
 		default:
@@ -950,3 +956,5 @@
 EXPORT_SYMBOL(snd_malloc_pages);
 EXPORT_SYMBOL(snd_malloc_pages_fallback);
 EXPORT_SYMBOL(snd_free_pages);
+
+#include "memalloc.inc1"
