--- ../alsa-kernel/core/misc.c	2009-08-27 17:44:33.000000000 +0200
+++ misc.c	2009-08-27 17:45:27.000000000 +0200
@@ -1,3 +1,9 @@
+#define __NO_VERSION__
+#include "adriver.h"
+#ifndef KERN_DEFAULT
+#define KERN_DEFAULT	""
+#endif
+
 /*
  *  Misc and compatibility things
  *  Copyright (c) by Jaroslav Kysela <perex@perex.cz>
@@ -42,7 +48,7 @@ void release_and_free_resource(struct re
 {
 	if (res) {
 		release_resource(res);
-		kfree(res);
+		kfree_nocheck(res);
 	}
 }
 
@@ -115,12 +121,20 @@ const struct snd_pci_quirk *
 snd_pci_quirk_lookup(struct pci_dev *pci, const struct snd_pci_quirk *list)
 {
 	const struct snd_pci_quirk *q;
+	unsigned short subsystem_vendor, subsystem_device;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 3, 0)
+	subsystem_vendor = pci->subsystem_vendor;
+	subsystem_device = pci->subsystem_device;
+#else
+	pci_read_config_word(pci, PCI_SUBSYSTEM_VENDOR_ID, &subsystem_vendor);
+	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &subsystem_device);
+#endif
 	for (q = list; q->subvendor; q++) {
-		if (q->subvendor != pci->subsystem_vendor)
+		if (q->subvendor != subsystem_vendor)
 			continue;
 		if (!q->subdevice ||
-		    (pci->subsystem_device & q->subdevice_mask) == q->subdevice)
+		    (subsystem_device & q->subdevice_mask) == q->subdevice)
 			return q;
 	}
 	return NULL;
