--- ../alsa-kernel/core/sound.c	2005-06-30 14:10:46.000000000 +0200
+++ sound.c	2005-06-30 12:46:16.000000000 +0200
@@ -64,7 +64,15 @@
 
 static DECLARE_MUTEX(sound_mutex);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
+#ifdef CONFIG_SND_HAVE_CLASS_SIMPLE
+extern struct class_simple *sound_class;
+#else
 extern struct class *sound_class;
+#endif
+#else
+#define sound_class NULL
+#endif
 
 
 #ifdef CONFIG_KMOD
@@ -160,7 +168,9 @@
 
 static struct file_operations snd_fops =
 {
+#ifndef LINUX_2_2
 	.owner =	THIS_MODULE,
+#endif
 	.open =		snd_open
 };
 
@@ -231,7 +241,11 @@
 		devfs_mk_cdev(MKDEV(major, minor), S_IFCHR | device_mode, "snd/%s", name);
 	if (card)
 		device = card->dev;
+#ifdef CONFIG_SND_HAVE_CLASS_SIMPLE
+	class_simple_device_add(sound_class, MKDEV(major, minor), device, name);
+#else
 	class_device_create(sound_class, MKDEV(major, minor), device, "%s", name);
+#endif
 
 	up(&sound_mutex);
 	return 0;
@@ -263,7 +277,11 @@
 
 	if (strncmp(mptr->name, "controlC", 8) || card->number >= cards_limit) /* created in sound.c */
 		devfs_remove("snd/%s", mptr->name);
+#ifdef CONFIG_SND_HAVE_CLASS_SIMPLE
+	class_simple_device_remove(MKDEV(major, minor));
+#else
 	class_device_destroy(sound_class, MKDEV(major, minor));
+#endif
 
 	list_del(&mptr->list);
 	up(&sound_mutex);
@@ -359,6 +377,9 @@
 #ifndef MODULE
 	printk(KERN_INFO "Advanced Linux Sound Architecture Driver Version " CONFIG_SND_VERSION CONFIG_SND_DATE ".\n");
 #endif
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,3,0) && defined(CONFIG_APM)
+	pm_init();
+#endif
 	return 0;
 }
 
@@ -371,6 +392,9 @@
 
 	snd_info_minor_unregister();
 	snd_info_done();
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,3,0) && defined(CONFIG_APM)
+	pm_done();
+#endif
 	snd_memory_done();
 	if (unregister_chrdev(major, "alsa") != 0)
 		snd_printk(KERN_ERR "unable to unregister major device number %d\n", major);
@@ -488,3 +512,5 @@
 EXPORT_SYMBOL(snd_wrapper_vmalloc);
 EXPORT_SYMBOL(snd_wrapper_vfree);
 #endif
+
+#include "sound.inc"
