--- awacs.c	2003-05-31 13:38:32.000000000 +0200
+++ awacs.c.old	2003-05-31 13:56:36.000000000 +0200
@@ -1,3 +1,10 @@
+#define __NO_VERSION__
+#include <linux/config.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,3,0)
+#define CONFIG_ADB_CUDA
+#define CONFIG_ADB_PMU
+#endif
 /*
  * PMac AWACS lowlevel functions
  *
@@ -32,7 +39,7 @@
 #define chip_t pmac_t
 
 
-#ifdef CONFIG_ADB_CUDA
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,3,0) || defined(CONFIG_ADB_CUDA)
 #define PMAC_AMP_AVAIL
 #endif
 
@@ -43,7 +50,11 @@
 	unsigned char amp_tone[2];
 } awacs_amp_t;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,3,0)
+#define CHECK_CUDA_AMP() (adb_hardware == ADB_VIACUDA)
+#else
 #define CHECK_CUDA_AMP() (sys_ctrler == SYS_CTRLER_CUDA)
+#endif
 
 #endif /* PMAC_AMP_AVAIL */
 
