--- ../../alsa-kernel/pci/korg1212/korg1212.c	2007-04-26 13:51:47.000000000 +0200
+++ korg1212.c	2007-04-26 14:08:42.000000000 +0200
@@ -2343,7 +2343,11 @@
 #ifdef CONFIG_SND_KORG1212_FIRMWARE_IN_KERNEL
 	dsp_code = &static_dsp_code;
 #else
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 5, 0)
 	err = request_firmware(&dsp_code, "korg/k1212.dsp", &pci->dev);
+#else
+	err = request_firmware(&dsp_code, "korg/k1212.dsp", pci_name(pci));
+#endif
 	if (err < 0) {
 		release_firmware(dsp_code);
 		snd_printk(KERN_ERR "firmware not available\n");
@@ -2500,3 +2504,4 @@
 
 module_init(alsa_card_korg1212_init)
 module_exit(alsa_card_korg1212_exit)
+EXPORT_NO_SYMBOLS;
