--- ../../alsa-kernel/pci/ali5451/ali5451.c	2007-07-19 16:36:25.000000000 +0200
+++ ali5451.c	2007-07-20 12:19:30.000000000 +0200
@@ -2217,7 +2217,7 @@
 	codec->card = card;
 	codec->pci = pci;
 	codec->irq = -1;
-	codec->revision = pci->revision;
+	codec->revision = snd_pci_revision(pci);
 	codec->spdif_support = spdif_support;
 
 	if (pcm_streams < 1)
@@ -2387,3 +2387,5 @@
 
 module_init(alsa_card_ali_init)
 module_exit(alsa_card_ali_exit)
+
+EXPORT_NO_SYMBOLS;
