--- ali5451.c	2004-04-07 16:58:56.596731870 +0200
+++ ali5451.c	2004-04-07 18:35:03.507430272 +0200
@@ -1977,6 +1977,7 @@
 	return;
 }
 
+#ifndef PCI_OLD_SUSPEND
 static int snd_ali_suspend(struct pci_dev *dev, u32 state)
 {
 	ali_t *chip = snd_magic_cast(ali_t, pci_get_drvdata(dev), return -ENXIO);
@@ -1989,6 +1990,18 @@
 	ali_resume(chip);
 	return 0;
 }
+#else
+static void snd_ali_suspend(struct pci_dev *dev)
+{
+	ali_t *chip = snd_magic_cast(ali_t, pci_get_drvdata(dev), return);
+	ali_suspend(chip);
+}
+static void snd_ali_resume(struct pci_dev *dev)
+{
+	ali_t *chip = snd_magic_cast(ali_t, pci_get_drvdata(dev), return);
+	ali_resume(chip);
+}
+#endif
 #endif /* CONFIG_PM */
 
 static int snd_ali_free(ali_t * codec)
@@ -2306,3 +2319,5 @@
 
 module_init(alsa_card_ali_init)
 module_exit(alsa_card_ali_exit)
+
+EXPORT_NO_SYMBOLS;
