--- ymfpci.c	2004-04-07 16:59:29.513264858 +0200
+++ ymfpci.c	2004-04-07 18:35:33.146723726 +0200
@@ -318,6 +318,7 @@
 }
 
 #ifdef CONFIG_PM
+#ifndef PCI_OLD_SUSPEND
 static int snd_card_ymfpci_suspend(struct pci_dev *pci, u32 state)
 {
 	ymfpci_t *chip = snd_magic_cast(ymfpci_t, pci_get_drvdata(pci), return -ENXIO);
@@ -330,6 +331,18 @@
 	snd_ymfpci_resume(chip);
 	return 0;
 }
+#else
+static void snd_card_ymfpci_suspend(struct pci_dev *pci)
+{
+	ymfpci_t *chip = snd_magic_cast(ymfpci_t, pci_get_drvdata(pci), return);
+	snd_ymfpci_suspend(chip);
+}
+static void snd_card_ymfpci_resume(struct pci_dev *pci)
+{
+	ymfpci_t *chip = snd_magic_cast(ymfpci_t, pci_get_drvdata(pci), return);
+	snd_ymfpci_resume(chip);
+}
+#endif
 #endif
 
 static void __devexit snd_card_ymfpci_remove(struct pci_dev *pci)
@@ -371,3 +384,5 @@
 
 module_init(alsa_card_ymfpci_init)
 module_exit(alsa_card_ymfpci_exit)
+
+EXPORT_NO_SYMBOLS;
