--- ../../alsa-kernel/pci/ctxfi/ctatc.c	2009-06-08 18:07:22.000000000 +0200
+++ ctatc.c	2009-06-08 18:15:00.000000000 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /**
  * Copyright (C) 2008, Creative Technology Ltd. All Rights Reserved.
  *
@@ -1239,10 +1240,12 @@
 		return -ENOENT;
 	atc->model = p->value;
 	atc->model_name = ct_subsys_name[atc->model];
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 	snd_printd("ctxfi: chip %s model %s (%04x:%04x) is found\n",
 		   atc->chip_name, atc->model_name,
 		   atc->pci->subsystem_vendor,
 		   atc->pci->subsystem_device);
+#endif
 	return 0;
 }
 
