--- ../../alsa-kernel/pci/asihpi/hpidspcd.c	2010-04-21 18:36:16.613453111 +0200
+++ hpidspcd.c	2010-04-21 18:41:26.104970371 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /***********************************************************************/
 /*!
 
@@ -73,8 +74,13 @@
 	sprintf(fw_name, "asihpi/dsp%04x.bin", adapter);
 	HPI_DEBUG_LOG(INFO, "requesting firmware for %s\n", fw_name);
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2 , 5 , 0))
+	err = request_firmware(&ps_firmware, fw_name,
+		ps_dsp_code->ps_dev->slot_name);
+#else
 	err = request_firmware(&ps_firmware, fw_name,
 		&ps_dsp_code->ps_dev->dev);
+#endif
 	if (err != 0) {
 		HPI_DEBUG_LOG(ERROR, "%d, request_firmware failed for  %s\n",
 			err, fw_name);
