--- cs46xx.c	2004-04-07 16:58:27.194401697 +0200
+++ cs46xx.c	2004-04-07 18:34:35.037872030 +0200
@@ -164,6 +164,7 @@
 }
 
 #ifdef CONFIG_PM
+#ifndef PCI_OLD_SUSPEND
 static int snd_card_cs46xx_suspend(struct pci_dev *pci, u32 state)
 {
 	cs46xx_t *chip = snd_magic_cast(cs46xx_t, pci_get_drvdata(pci), return -ENXIO);
@@ -176,6 +177,18 @@
 	snd_cs46xx_resume(chip);
 	return 0;
 }
+#else
+static void snd_card_cs46xx_suspend(struct pci_dev *pci)
+{
+	cs46xx_t *chip = snd_magic_cast(cs46xx_t, pci_get_drvdata(pci), return);
+	snd_cs46xx_suspend(chip);
+}
+static void snd_card_cs46xx_resume(struct pci_dev *pci)
+{
+	cs46xx_t *chip = snd_magic_cast(cs46xx_t, pci_get_drvdata(pci), return);
+	snd_cs46xx_resume(chip);
+}
+#endif
 #endif
 
 static void __devexit snd_card_cs46xx_remove(struct pci_dev *pci)
@@ -217,3 +230,5 @@
 
 module_init(alsa_card_cs46xx_init)
 module_exit(alsa_card_cs46xx_exit)
+
+EXPORT_NO_SYMBOLS;
