--- ../../../alsa-kernel/aoa/soundbus/i2sbus/core.c	2011-10-19 17:32:44.000000000 +0200
+++ core.c	2012-07-23 11:42:28.000000000 +0200
@@ -1,3 +1,4 @@
+#include "ppc-prom-hack.h"
 /*
  * i2sbus driver
  *
@@ -221,9 +222,15 @@
 
 	mutex_init(&dev->lock);
 	spin_lock_init(&dev->low_lock);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 	dev->sound.ofdev.archdata.dma_mask = macio->ofdev.archdata.dma_mask;
 	dev->sound.ofdev.dev.of_node = np;
 	dev->sound.ofdev.dev.dma_mask = &dev->sound.ofdev.archdata.dma_mask;
+#else
+	dev->sound.ofdev.dma_mask = macio->ofdev.dma_mask;
+	dev->sound.ofdev.node = np;
+	dev->sound.ofdev.dev.dma_mask = &dev->sound.ofdev.dma_mask;
+#endif
 	dev->sound.ofdev.dev.parent = &macio->ofdev.dev;
 	dev->sound.ofdev.dev.release = i2sbus_release_dev;
 	dev->sound.attach_codec = i2sbus_attach_codec;
@@ -345,7 +352,11 @@
 		return -ENODEV;
 	}
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 	while ((np = of_get_next_child(dev->ofdev.dev.of_node, np))) {
+#else
+	while ((np = of_get_next_child(dev->ofdev.node, np))) {
+#endif
 		if (of_device_is_compatible(np, "i2sbus") ||
 		    of_device_is_compatible(np, "i2s-modem")) {
 			got += i2sbus_add_dev(dev, control, np);
@@ -436,11 +447,17 @@
 }
 
 static struct macio_driver i2sbus_drv = {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 	.driver = {
 		.name = "soundbus-i2s",
 		.owner = THIS_MODULE,
 		.of_match_table = i2sbus_match,
 	},
+#else
+	.name = "soundbus-i2s",
+	.owner = THIS_MODULE,
+	.match_table = i2sbus_match,
+#endif
 	.probe = i2sbus_probe,
 	.remove = i2sbus_remove,
 #ifdef CONFIG_PM
