--- ../../alsa-kernel/usb/usx2y/usbusx2yaudio.c	Tue Oct 26 20:38:21 2004
+++ usbusx2yaudio.c	Tue Oct 26 20:39:31 2004
@@ -299,7 +310,11 @@
 	usX2Y_clients_stop(usX2Y);
 }
 
+#ifndef OLD_USB
 static void i_usX2Y_urb_complete(struct urb *urb, struct pt_regs *regs)
+#else
+static void i_usX2Y_urb_complete(struct urb *urb)
+#endif
 {
 	snd_usX2Y_substream_t *subs = (snd_usX2Y_substream_t*)urb->context;
 	usX2Ydev_t *usX2Y = subs->usX2Y;
@@ -337,7 +352,11 @@
 	}
 }
 
+#ifndef OLD_USB
 static void usX2Y_urbs_set_complete(usX2Ydev_t * usX2Y, void (*complete)(struct urb *, struct pt_regs *))
+#else
+static void usX2Y_urbs_set_complete(usX2Ydev_t * usX2Y, void (*complete)(struct urb *))
+#endif
 {
 	int s, u;
 	for (s = 0; s < 4; s++) {
@@ -357,7 +376,11 @@
 	usX2Y->prepare_subs = NULL;
 }
 
+#ifndef OLD_USB
 static void i_usX2Y_subs_startup(struct urb *urb, struct pt_regs *regs)
+#else
+static void i_usX2Y_subs_startup(struct urb *urb)
+#endif
 {
 	snd_usX2Y_substream_t *subs = (snd_usX2Y_substream_t*)urb->context;
 	usX2Ydev_t *usX2Y = subs->usX2Y;
@@ -369,7 +392,11 @@
 			wake_up(&usX2Y->prepare_wait_queue);
 		}
 
+#ifndef OLD_USB
 	i_usX2Y_urb_complete(urb, regs);
+#else
+	i_usX2Y_urb_complete(urb);
+#endif
 }
 
 static void usX2Y_subs_prepare(snd_usX2Y_substream_t *subs)
@@ -654,7 +681,11 @@
 };
 #define NOOF_SETRATE_URBS ARRAY_SIZE(SetRate48000)
 
+#ifndef OLD_USB
 static void i_usX2Y_04Int(struct urb* urb, struct pt_regs *regs)
+#else
+static void i_usX2Y_04Int(struct urb* urb)
+#endif
 {
 	usX2Ydev_t*	usX2Y = urb->context;
 	
