--- ../../alsa-kernel/pci/ca0106/ca0106_main.c	2007-07-19 16:36:25.000000000 +0200
+++ ca0106_main.c	2007-07-20 12:46:22.000000000 +0200
@@ -165,7 +165,7 @@
 module_param_array(subsystem, uint, NULL, 0444);
 MODULE_PARM_DESC(subsystem, "Force card subsystem model.");
 
-#include "ca0106.h"
+#include "../../alsa-kernel/pci/ca0106/ca0106.h"
 
 static struct snd_ca0106_details ca0106_chip_details[] = {
 	 /* Sound Blaster X-Fi Extreme Audio. This does not have an AC97. 53SB079000000 */
@@ -1317,7 +1317,7 @@
 	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &chip->model);
 #if 1
 	printk(KERN_INFO "snd-ca0106: Model %04x Rev %08x Serial %08x\n", chip->model,
-	       pci->revision, chip->serial);
+	       snd_pci_revision(pci), chip->serial);
 #endif
 	strcpy(card->driver, "CA0106");
 	strcpy(card->shortname, "CA0106");
@@ -1681,3 +1681,5 @@
 
 module_init(alsa_card_ca0106_init)
 module_exit(alsa_card_ca0106_exit)
+
+EXPORT_NO_SYMBOLS;
