--- cs4281.c	2003-06-10 15:21:01.000000000 +0200
+++ cs4281.c.old	2003-06-10 15:20:49.000000000 +0200
@@ -2119,6 +2119,7 @@
 	snd_power_change_state(card, SNDRV_CTL_POWER_D0);
 }
 
+#ifndef PCI_OLD_SUSPEND
 static int snd_cs4281_suspend(struct pci_dev *dev, u32 state)
 {
 	cs4281_t *chip = snd_magic_cast(cs4281_t, pci_get_drvdata(dev), return -ENXIO);
@@ -2131,6 +2132,18 @@
 	cs4281_resume(chip);
 	return 0;
 }
+#else
+static void snd_cs4281_suspend(struct pci_dev *dev)
+{
+	cs4281_t *chip = snd_magic_cast(cs4281_t, pci_get_drvdata(dev), return);
+	cs4281_suspend(chip);
+}
+static void snd_cs4281_resume(struct pci_dev *dev)
+{
+	cs4281_t *chip = snd_magic_cast(cs4281_t, pci_get_drvdata(dev), return);
+	cs4281_resume(chip);
+}
+#endif
 
 /* callback */
 static int snd_cs4281_set_power_state(snd_card_t *card, unsigned int power_state)
@@ -2206,3 +2219,4 @@
 __setup("snd-cs4281=", alsa_card_cs4281_setup);
 
 #endif /* ifndef MODULE */
+EXPORT_NO_SYMBOLS;
