--- es1968.c	2004-04-07 16:58:13.077604048 +0200
+++ es1968.c	2004-04-07 18:33:54.181116422 +0200
@@ -2463,6 +2463,7 @@
 	snd_power_change_state(card, SNDRV_CTL_POWER_D0);
 }
 
+#ifndef PCI_OLD_SUSPEND
 static int snd_es1968_suspend(struct pci_dev *dev, u32 state)
 {
 	es1968_t *chip = snd_magic_cast(es1968_t, pci_get_drvdata(dev), return -ENXIO);
@@ -2475,6 +2476,18 @@
 	es1968_resume(chip);
 	return 0;
 }
+#else
+static void snd_es1968_suspend(struct pci_dev *dev)
+{
+	es1968_t *chip = snd_magic_cast(es1968_t, pci_get_drvdata(dev), return);
+	es1968_suspend(chip);
+}
+static void snd_es1968_resume(struct pci_dev *dev)
+{
+	es1968_t *chip = snd_magic_cast(es1968_t, pci_get_drvdata(dev), return);
+	es1968_resume(chip);
+}
+#endif
 
 /* callback */
 static int snd_es1968_set_power_state(snd_card_t *card, unsigned int power_state)
@@ -2811,3 +2824,5 @@
 
 module_init(alsa_card_es1968_init)
 module_exit(alsa_card_es1968_exit)
+
+EXPORT_NO_SYMBOLS;
