--- es1968.c	2003-06-10 15:20:37.000000000 +0200
+++ es1968.c.old	2003-06-10 15:20:22.000000000 +0200
@@ -2468,6 +2468,7 @@
 	snd_power_change_state(card, SNDRV_CTL_POWER_D0);
 }
 
+#ifndef PCI_OLD_SUSPEND
 static int snd_es1968_suspend(struct pci_dev *dev, u32 state)
 {
 	es1968_t *chip = snd_magic_cast(es1968_t, pci_get_drvdata(dev), return -ENXIO);
@@ -2480,6 +2481,18 @@
 	es1968_resume(chip);
 	return 0;
 }
+#else
+static void snd_es1968_suspend(struct pci_dev *dev)
+{
+	es1968_t *chip = snd_magic_cast(es1968_t, pci_get_drvdata(dev), return);
+	es1968_suspend(chip);
+}
+static void snd_es1968_resume(struct pci_dev *dev)
+{
+	es1968_t *chip = snd_magic_cast(es1968_t, pci_get_drvdata(dev), return);
+	es1968_resume(chip);
+}
+#endif
 
 /* callback */
 static int snd_es1968_set_power_state(snd_card_t *card, unsigned int power_state)
@@ -2868,3 +2881,4 @@
 __setup("snd-es1968=", alsa_card_es1968_setup);
 
 #endif /* ifndef MODULE */
+EXPORT_NO_SYMBOLS;
