--- intel8x0.c	2003-06-10 15:08:31.000000000 +0200
+++ intel8x0.c.old	2003-06-10 15:10:32.000000000 +0200
@@ -1942,6 +1942,7 @@
 	snd_power_change_state(card, SNDRV_CTL_POWER_D0);
 }
 
+#ifndef PCI_OLD_SUSPEND
 static int snd_intel8x0_suspend(struct pci_dev *dev, u32 state)
 {
 	intel8x0_t *chip = snd_magic_cast(intel8x0_t, pci_get_drvdata(dev), return -ENXIO);
@@ -1954,6 +1955,18 @@
 	intel8x0_resume(chip);
 	return 0;
 }
+#else
+static void snd_intel8x0_suspend(struct pci_dev *dev)
+{
+	intel8x0_t *chip = snd_magic_cast(intel8x0_t, pci_get_drvdata(dev), return);
+	intel8x0_suspend(chip);
+}
+static void snd_intel8x0_resume(struct pci_dev *dev)
+{
+	intel8x0_t *chip = snd_magic_cast(intel8x0_t, pci_get_drvdata(dev), return);
+	intel8x0_resume(chip);
+}
+#endif
 
 /* callback */
 static int snd_intel8x0_set_power_state(snd_card_t *card, unsigned int power_state)
@@ -2539,3 +2552,4 @@
 __setup("snd-intel8x0=", alsa_card_intel8x0_setup);
 
 #endif /* ifndef MODULE */
+EXPORT_NO_SYMBOLS;
