--- ../alsa-kernel/pci/maestro3.c	2006-11-02 22:50:09.000000000 +0100
+++ maestro3.c	2006-11-02 22:54:12.000000000 +0100
@@ -2710,6 +2710,7 @@
 	int i, err;
 	const struct m3_quirk *quirk;
 	const struct m3_hv_quirk *hv_quirk;
+	u16 subsystem_vendor, subsystem_device;
 	static struct snd_device_ops ops = {
 		.dev_free =	snd_m3_dev_free,
 	};
@@ -2733,6 +2734,14 @@
 		return -ENOMEM;
 	}
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 3, 0)
+	subsystem_vendor = pci->subsystem_vendor;
+	subsystem_device = pci->subsystem_device;
+#else
+	pci_read_config_word(pci, PCI_SUBSYSTEM_VENDOR_ID, &subsystem_vendor);
+	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &subsystem_device);
+#endif
+
 	spin_lock_init(&chip->reg_lock);
 	spin_lock_init(&chip->ac97_lock);
 
@@ -2750,8 +2759,8 @@
 	chip->irq = -1;
 
 	for (quirk = m3_quirk_list; quirk->vendor; quirk++) {
-		if (pci->subsystem_vendor == quirk->vendor &&
-		    pci->subsystem_device == quirk->device) {
+		if (subsystem_vendor == quirk->vendor &&
+		    subsystem_device == quirk->device) {
 			printk(KERN_INFO "maestro3: enabled hack for '%s'\n", quirk->name);
 			chip->quirk = quirk;
 			break;
@@ -2761,8 +2770,8 @@
 	for (hv_quirk = m3_hv_quirk_list; hv_quirk->vendor; hv_quirk++) {
 		if (pci->vendor == hv_quirk->vendor &&
 		    pci->device == hv_quirk->device &&
-		    pci->subsystem_vendor == hv_quirk->subsystem_vendor &&
-		    pci->subsystem_device == hv_quirk->subsystem_device) {
+		    subsystem_vendor == hv_quirk->subsystem_vendor &&
+		    subsystem_device == hv_quirk->subsystem_device) {
 			chip->hv_quirk = hv_quirk;
 			break;
 		}
@@ -2787,8 +2796,13 @@
 		return -ENOMEM;
 	}
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 5, 0)
 	err = request_firmware(&chip->assp_kernel_image,
 			       "ess/maestro3_assp_kernel.fw", &pci->dev);
+#else
+	err = request_firmware(&chip->assp_kernel_image,
+			       "ess/maestro3_assp_kernel.fw", pci_name(pci));
+#endif
 	if (err < 0) {
 #ifdef FIRMWARE_IN_THE_KERNEL
 		chip->assp_kernel_image = &assp_kernel;
@@ -2799,8 +2813,13 @@
 	} else
 		snd_m3_convert_from_le(chip->assp_kernel_image);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 5, 0)
 	err = request_firmware(&chip->assp_minisrc_image,
 			       "ess/maestro3_assp_minisrc.fw", &pci->dev);
+#else
+	err = request_firmware(&chip->assp_minisrc_image,
+			       "ess/maestro3_assp_minisrc.fw", pci_name(pci));
+#endif
 	if (err < 0) {
 #ifdef FIRMWARE_IN_THE_KERNEL
 		chip->assp_minisrc_image = &assp_minisrc;
@@ -2971,3 +2990,5 @@
 
 module_init(alsa_card_m3_init)
 module_exit(alsa_card_m3_exit)
+
+EXPORT_NO_SYMBOLS;
