--- ../alsa-kernel/pci/via82xx_modem.c	2012-04-24 12:18:48.000000000 +0200
+++ via82xx_modem.c	2012-04-24 12:33:18.000000000 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  *   ALSA modem driver for VIA VT82xx (South Bridge)
  *
@@ -1186,7 +1187,7 @@
 		goto __error;
 	}
 		
-	if ((err = snd_via82xx_create(card, pci, chip_type, pci->revision,
+	if ((err = snd_via82xx_create(card, pci, chip_type, snd_pci_revision(pci),
 				      ac97_clock, &chip)) < 0)
 		goto __error;
 	card->private_data = chip;
@@ -1235,3 +1236,5 @@
 };
 
 module_pci_driver(via82xx_modem_driver);
+
+EXPORT_NO_SYMBOLS;
