--- nm256.c	2003-06-10 15:13:55.000000000 +0200
+++ nm256.c.old	2003-06-10 15:14:45.000000000 +0200
@@ -1307,6 +1307,7 @@
 	snd_power_change_state(card, SNDRV_CTL_POWER_D0);
 }
 
+#ifndef PCI_OLD_SUSPEND
 static int snd_nm256_suspend(struct pci_dev *dev, u32 state)
 {
 	nm256_t *chip = snd_magic_cast(nm256_t, pci_get_drvdata(dev), return -ENXIO);
@@ -1319,6 +1320,18 @@
 	nm256_resume(chip);
 	return 0;
 }
+#else
+static void snd_nm256_suspend(struct pci_dev *dev)
+{
+	nm256_t *chip = snd_magic_cast(nm256_t, pci_get_drvdata(dev), return);
+	nm256_suspend(chip);
+}
+static void snd_nm256_resume(struct pci_dev *dev)
+{
+	nm256_t *chip = snd_magic_cast(nm256_t, pci_get_drvdata(dev), return);
+	nm256_resume(chip);
+}
+#endif
 
 /* callback */
 static int snd_nm256_set_power_state(snd_card_t *card, unsigned int power_state)
@@ -1707,3 +1720,4 @@
 __setup("snd-nm256=", alsa_card_nm256_setup);
 
 #endif /* ifndef MODULE */
+EXPORT_NO_SYMBOLS;
