--- ../../alsa-kernel/pci/hda/hda_codec.c	2006-12-19 17:00:25.000000000 +0100
+++ hda_codec.c	2006-12-19 17:04:25.000000000 +0100
@@ -263,7 +263,10 @@
 	unsol->queue[wp] = res;
 	unsol->queue[wp + 1] = res_ex;
 
+/* 2.4 kernels seem to have a problem with workq wrapper... */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 	schedule_work(&unsol->work);
+#endif
 
 	return 0;
 }
@@ -300,6 +303,7 @@
  */
 static int init_unsol_queue(struct hda_bus *bus)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 	struct hda_bus_unsolicited *unsol;
 
 	if (bus->unsol) /* already initialized */
@@ -313,6 +317,7 @@
 	INIT_WORK(&unsol->work, process_unsol_events);
 	unsol->bus = bus;
 	bus->unsol = unsol;
+#endif
 	return 0;
 }
 
@@ -328,7 +333,9 @@
 	if (! bus)
 		return 0;
 	if (bus->unsol) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 		flush_scheduled_work();
+#endif
 		kfree(bus->unsol);
 	}
 	list_for_each_safe(p, n, &bus->codec_list) {
