--- ../../alsa-kernel/pci/hda/hda_codec.c	2007-04-27 21:41:42.000000000 +0200
+++ hda_codec.c	2007-04-30 09:14:49.000000000 +0200
@@ -1,3 +1,4 @@
+#define __NO_VERSION__
 /*
  * Universal Interface for Intel High Definition Audio Codec
  *
@@ -256,7 +257,10 @@
 	unsol->queue[wp] = res;
 	unsol->queue[wp + 1] = res_ex;
 
+/* 2.4 kernels seem to have a problem with workq wrapper... */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 	schedule_work(&unsol->work);
+#endif
 
 	return 0;
 }
@@ -291,6 +295,7 @@
  */
 static int __devinit init_unsol_queue(struct hda_bus *bus)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 	struct hda_bus_unsolicited *unsol;
 
 	if (bus->unsol) /* already initialized */
@@ -305,6 +310,7 @@
 	INIT_WORK(&unsol->work, process_unsol_events);
 	unsol->bus = bus;
 	bus->unsol = unsol;
+#endif
 	return 0;
 }
 
@@ -320,7 +326,9 @@
 	if (!bus)
 		return 0;
 	if (bus->unsol) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 		flush_scheduled_work();
+#endif
 		kfree(bus->unsol);
 	}
 	list_for_each_entry_safe(codec, n, &bus->codec_list, list) {
