--- ../../alsa-kernel/pcmcia/vx/vxpocket.c	2004-11-08 12:48:27.000000000 +0100
+++ vxpocket.c	2004-11-17 14:48:22.458415985 +0100
@@ -35,6 +35,7 @@
 #include <linux/moduleparam.h>
 #include <sound/core.h>
 #include <pcmcia/version.h>
+#include "compat_cs.h"
 #include "vxpocket.h"
 #include <sound/initval.h>
 
@@ -170,3 +175,5 @@
 
 module_init(init_vxpocket);
 module_exit(exit_vxpocket);
+
+EXPORT_NO_SYMBOLS;
