--- ../../alsa-kernel/pcmcia/vx/vxpocket.c	2005-06-30 12:13:27.000000000 +0200
+++ vxpocket.c	2005-06-30 12:13:26.000000000 +0200
@@ -24,6 +24,7 @@
 #include <linux/moduleparam.h>
 #include <sound/core.h>
 #include <pcmcia/version.h>
+#include "compat_cs.h"
 #include "vxpocket.h"
 #include <sound/initval.h>
 
@@ -129,11 +130,13 @@
  * Module entry points
  */
 
+#ifdef PCMCIA_DEVICE_MANF_CARD
 static struct pcmcia_device_id vxp_ids[] = {
 	PCMCIA_DEVICE_MANF_CARD(0x01f1, 0x0100),
 	PCMCIA_DEVICE_NULL
 };
 MODULE_DEVICE_TABLE(pcmcia, vxp_ids);
+#endif
 
 static struct pcmcia_driver vxp_cs_driver = {
 	.owner		= THIS_MODULE,
@@ -142,7 +145,9 @@
 	},
 	.attach		= vxp_attach,
 	.detach		= vxp_detach,
+#ifdef PCMCIA_DEVICE_MANF_CARD
 	.id_table	= vxp_ids,
+#endif
 };
 
 static int __init init_vxpocket(void)
@@ -158,3 +163,5 @@
 
 module_init(init_vxpocket);
 module_exit(exit_vxpocket);
+
+EXPORT_NO_SYMBOLS;
