--- ../../alsa-kernel/aoa/soundbus/core.c	2011-10-19 17:32:44.000000000 +0200
+++ core.c	2012-07-20 16:57:25.000000000 +0200
@@ -1,3 +1,4 @@
+#include "ppc-prom-hack.h"
 /*
  * soundbus
  *
@@ -74,11 +75,19 @@
 	of = &soundbus_dev->ofdev;
 
 	/* stuff we want to pass to /sbin/hotplug */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 	retval = add_uevent_var(env, "OF_NAME=%s", of->dev.of_node->name);
+#else
+	retval = add_uevent_var(env, "OF_NAME=%s", of->node->name);
+#endif
 	if (retval)
 		return retval;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 	retval = add_uevent_var(env, "OF_TYPE=%s", of->dev.of_node->type);
+#else
+	retval = add_uevent_var(env, "OF_TYPE=%s", of->node->type);
+#endif
 	if (retval)
 		return retval;
 
@@ -86,7 +95,11 @@
 	 * it's not really legal to split it out with commas. We split it
 	 * up using a number of environment variables instead. */
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 	compat = of_get_property(of->dev.of_node, "compatible", &cplen);
+#else
+	compat = of_get_property(of->node, "compatible", &cplen);
+#endif
 	while (compat && cplen > 0) {
 		int tmp = env->buflen;
 		retval = add_uevent_var(env, "OF_COMPATIBLE_%d=%s", seen, compat);
@@ -169,7 +182,11 @@
 
 	/* sanity checks */
 	if (!dev->attach_codec ||
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 	    !dev->ofdev.dev.of_node ||
+#else
+	    !dev->ofdev.node ||
+#endif
 	    dev->pcmname ||
 	    dev->pcmid != -1) {
 		printk(KERN_ERR "soundbus: adding device failed sanity check!\n");
