--- ../../alsa-kernel/pci/emu10k1/emu10k1x.c	2012-04-24 12:18:48.000000000 +0200
+++ emu10k1x.c	2012-04-24 12:35:32.000000000 +0200
@@ -1,3 +1,4 @@
+#include "adriver.h"
 /*
  *  Copyright (c) by Francisco Moraes <fmoraes@nc.rr.com>
  *  Driver EMU10K1X chips
@@ -940,7 +941,7 @@
 
 	pci_set_master(pci);
 	/* read revision & serial */
-	chip->revision = pci->revision;
+	chip->revision = snd_pci_revision(pci);
 	pci_read_config_dword(pci, PCI_SUBSYSTEM_VENDOR_ID, &chip->serial);
 	pci_read_config_word(pci, PCI_SUBSYSTEM_ID, &chip->model);
 	snd_printk(KERN_INFO "Model %04x Rev %08x Serial %08x\n", chip->model,
@@ -1620,3 +1621,5 @@
 };
 
 module_pci_driver(emu10k1x_driver);
+
+EXPORT_NO_SYMBOLS;
