--- ../../alsa-kernel/usb/caiaq/audio.c	2013-03-04 09:58:34.547644219 +0100
+++ audio.c	2013-03-04 10:03:35.120836851 +0100
@@ -1,3 +1,5 @@
+#include "adriver.h"
+#include "../usbcompat.h"
 /*
  *   Copyright (c) 2006-2008 Daniel Mack, Karsten Wiese
  *
@@ -626,7 +628,11 @@
 	}
 }
 
+#if !defined(OLD_USB) && !defined(CONFIG_SND_NEW_IRQ_HANDLER)
+static void read_completed(struct urb *urb, struct pt_regs *regs)
+#else
 static void read_completed(struct urb *urb)
+#endif
 {
 	struct snd_usb_caiaq_cb_info *info = urb->context;
 	struct snd_usb_caiaqdev *cdev;
@@ -703,7 +709,11 @@
 	usb_submit_urb(urb, GFP_ATOMIC);
 }
 
+#if !defined(OLD_USB) && !defined(CONFIG_SND_NEW_IRQ_HANDLER)
+static void write_completed(struct urb *urb, struct pt_regs *regs)
+#else
 static void write_completed(struct urb *urb)
+#endif
 {
 	struct snd_usb_caiaq_cb_info *info = urb->context;
 	struct snd_usb_caiaqdev *cdev = info->cdev;
