--- trident.c	2003-06-10 15:25:34.000000000 +0200
+++ trident.c.old	2003-06-10 15:25:09.000000000 +0200
@@ -182,6 +182,7 @@
 }
 
 #ifdef CONFIG_PM
+#ifndef PCI_OLD_SUSPEND
 static int snd_card_trident_suspend(struct pci_dev *pci, u32 state)
 {
 	trident_t *chip = snd_magic_cast(trident_t, pci_get_drvdata(pci), return -ENXIO);
@@ -194,6 +195,18 @@
 	snd_trident_resume(chip);
 	return 0;
 }
+#else
+static void snd_card_trident_suspend(struct pci_dev *pci)
+{
+	trident_t *chip = snd_magic_cast(trident_t, pci_get_drvdata(pci), return);
+	snd_trident_suspend(chip);
+}
+static void snd_card_trident_resume(struct pci_dev *pci)
+{
+	trident_t *chip = snd_magic_cast(trident_t, pci_get_drvdata(pci), return);
+	snd_trident_resume(chip);
+}
+#endif
 #endif
 
 static struct pci_driver driver = {
