--- ../alsa-kernel/usb/usbmidi.c	2005-04-30 10:42:02.000000000 +0200
+++ usbmidi.c	2005-04-30 10:42:11.000000000 +0200
@@ -1,3 +1,5 @@
+#include "usbmidi.inc"
+
 /*
  * usbmidi.c - ALSA USB MIDI driver
  *
@@ -286,8 +288,13 @@
 		return -ENOMEM;
 	memcpy(buf, data, len);
 	dump_urb("sending", buf, len);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 12)
 	err = usb_bulk_msg(ep->umidi->chip->dev, ep->urb->pipe, buf, len,
 			   NULL, 1000);
+#else
+	err = usb_bulk_msg(ep->umidi->chip->dev, ep->urb->pipe, buf, len,
+			   NULL, HZ);
+#endif
 	kfree(buf);
 	return err;
 }
@@ -1230,7 +1237,11 @@
 	intf = umidi->iface;
 	if (!intf || intf->num_altsetting < 1)
 		return -ENOENT;
+#ifndef OLD_USB
 	hostif = intf->cur_altsetting;
+#else
+	hostif = &intf->altsetting[intf->act_altsetting];
+#endif
 	intfd = get_iface_desc(hostif);
 
 	for (i = 0; i < intfd->bNumEndpoints; ++i) {
@@ -1560,3 +1571,5 @@
 EXPORT_SYMBOL(snd_usbmidi_input_stop);
 EXPORT_SYMBOL(snd_usbmidi_input_start);
 EXPORT_SYMBOL(snd_usbmidi_disconnect);
+
+#include "usbmidi.inc1"
