--- ../alsa-kernel/soc/soc-core.c	2006-12-15 09:29:12.000000000 +0100
+++ soc-core.c	2006-12-15 16:21:22.000000000 +0100
@@ -730,7 +730,11 @@
 static void close_delayed_work(struct work_struct *work)
 {
 	struct snd_soc_device *socdev =
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 20)
 		container_of(work, struct snd_soc_device, delayed_work.work);
+#else
+		container_of(work, struct snd_soc_device, delayed_work.data);
+#endif
 	struct snd_soc_codec *codec = socdev->codec;
 	struct snd_soc_codec_dai *codec_dai;
 	int i;
