--- ../alsa-kernel/soc/soc-core.c	2010-09-30 23:00:05.764410616 +0200
+++ soc-core.c	2010-10-04 09:16:06.920758499 +0200
@@ -1,3 +1,8 @@
+#include "adriver.h"
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 27)
+/* HACK HACK - debugfs_remove_recursive() isn't defined */
+#undef CONFIG_DEBUG_FS
+#endif
 /*
  * soc-core.c  --  ALSA SoC Audio Layer
  *
@@ -1452,7 +1457,9 @@
 
 	/* register the rtd device */
 	rtd->dev.release = rtd_release;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
 	rtd->dev.init_name = dai_link->name;
+#endif
 	ret = device_register(&rtd->dev);
 	if (ret < 0) {
 		printk(KERN_ERR "asoc: failed to register DAI runtime device %d\n", ret);
@@ -1693,21 +1700,43 @@
 	return 0;
 }
 
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 30)
 static const struct dev_pm_ops soc_pm_ops = {
 	.suspend = soc_suspend,
 	.resume = soc_resume,
 	.poweroff = soc_poweroff,
 };
+#else
+#ifdef CONFIG_PM
+static int soc_old_suspend(struct platform_device *pdev, pm_message_t state)
+{
+	return soc_suspend(&pdev->dev);
+}
+static int soc_old_resume(struct platform_device *pdev)
+{
+	return soc_resume(&pdev->dev);
+}
+#else
+#define soc_old_suspend		NULL
+#define soc_old_resume		NULL
+#endif
+#endif /* < 2.6.30 */
 
 /* ASoC platform driver */
 static struct platform_driver soc_driver = {
 	.driver		= {
 		.name		= "soc-audio",
 		.owner		= THIS_MODULE,
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 30)
 		.pm		= &soc_pm_ops,
+#endif
 	},
 	.probe		= soc_probe,
 	.remove		= soc_remove,
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(2, 6, 30)
+	.suspend = soc_old_suspend,
+	.resume = soc_old_resume,
+#endif
 };
 
 /* create a new pcm */
