--- ../../alsa-kernel/drivers/mpu401/mpu401.c	2005-02-05 12:39:45.000000000 +0100
+++ mpu401.c	2005-02-05 12:50:41.000000000 +0100
@@ -28,6 +28,10 @@
 #include <sound/mpu401.h>
 #include <sound/initval.h>
 
+#ifndef IORESOURCE_DISABLED
+#define IORESOURCE_DISABLED 0	/* ugly hack */
+#endif
+
 MODULE_AUTHOR("Jaroslav Kysela <perex@suse.cz>");
 MODULE_DESCRIPTION("MPU-401 UART");
 MODULE_LICENSE("GPL");
@@ -40,6 +44,9 @@
 #endif
 static long port[SNDRV_CARDS] = SNDRV_DEFAULT_PORT;	/* MPU-401 port number */
 static int irq[SNDRV_CARDS] = SNDRV_DEFAULT_IRQ;	/* MPU-401 IRQ */
+#ifdef CONFIG_X86_PC9800
+static int pc98ii[SNDRV_CARDS];				/* PC98-II dauther board */
+#endif
 
 module_param_array(index, int, NULL, 0444);
 MODULE_PARM_DESC(index, "Index value for MPU-401 device.");
@@ -55,6 +62,10 @@
 MODULE_PARM_DESC(port, "Port # for MPU-401 device.");
 module_param_array(irq, int, NULL, 0444);
 MODULE_PARM_DESC(irq, "IRQ # for MPU-401 device.");
+#ifdef CONFIG_X86_PC9800
+module_param_array(pc98ii, bool, boot_devs, 0444);
+MODULE_PARM_DESC(pc98ii, "Roland MPU-PC98II support.");
+#endif
 
 static snd_card_t *snd_mpu401_legacy_cards[SNDRV_CARDS] = SNDRV_DEFAULT_PTR;
 
@@ -136,6 +147,9 @@
 	}
 #endif
 	if ((err = snd_mpu401_uart_new(card, 0,
+#ifdef CONFIG_X86_PC9800
+				       pc98ii[dev] ? MPU401_HW_PC98II :
+#endif
 				       MPU401_HW_MPU401,
 				       port[dev], 0,
 				       irq[dev], irq[dev] >= 0 ? SA_INTERRUPT : 0, NULL)) < 0) {
@@ -233,3 +247,5 @@
 
 module_init(alsa_card_mpu401_init)
 module_exit(alsa_card_mpu401_exit)
+
+EXPORT_NO_SYMBOLS;
