--- ../../alsa-kernel/pci/ac97/ac97_codec.c	2006-12-14 14:08:48.000000000 +0100
+++ ac97_codec.c	2006-12-15 15:37:14.000000000 +0100
@@ -35,9 +35,9 @@
 #include <sound/ac97_codec.h>
 #include <sound/asoundef.h>
 #include <sound/initval.h>
-#include "ac97_local.h"
-#include "ac97_id.h"
-#include "ac97_patch.h"
+#include "../../alsa-kernel/pci/ac97/ac97_local.h"
+#include "../../alsa-kernel/pci/ac97/ac97_id.h"
+#include "../../alsa-kernel/pci/ac97/ac97_patch.h"
 
 MODULE_AUTHOR("Jaroslav Kysela <perex@suse.cz>");
 MODULE_DESCRIPTION("Universal interface for Audio Codec '97");
@@ -1895,6 +1895,7 @@
 
 EXPORT_SYMBOL(snd_ac97_bus);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 /* stop no dev release warning */
 static void ac97_device_release(struct device * dev)
 {
@@ -1928,6 +1929,10 @@
 		device_unregister(&ac97->dev);
 	return 0;
 }
+#else /* <= 2.6.0 */
+#define snd_ac97_dev_register	NULL
+#define snd_ac97_dev_disconnect	NULL
+#endif
 
 /* build_ops to do nothing */
 static struct snd_ac97_build_ops null_build_ops;
